### PR TITLE
feat: implement resource monitor

### LINKED ISSUE #158: add resource alerts
e.g. disk >80%, low memory, etc

Investigate whether fly.io allows shipping such metrics to datadog, otherwise create a lightweight in-process subsystem that samples and logs them at WARN level, and create alerts in datadog.

### PR DESCRIPTION
**Pull request by boudra**
closes #158 

this is a resource monitor that polls for metrics and logs them, at the moment only inodes and disk space are logged

- it reports metrics every minute via logging, the idea is that we can use datadog alerts when those values are over a specific threshold
- there's a flag to disable it in case it gets noisy in dev or someone is in a platform where it doesn't work, like windows
- ended up going with: https://www.npmjs.com/package/diskstats
    - plenty of ways to get disk usage but inodes are trickier
    - found out that it was not possible to use inode reporting from `/proc` because: https://serverfault.com/a/970997
    - was going to roll my own df parser :D but found this library, it also relies on parsing `df` output for the metrics but they have tests for it
- [x] check if datadog supports alerting just from usage / total, or do we have to log a percentage (edit: it does, their metrics and monitoring tools are pretty advanced)
- [x] test if it works in Linux, I am thinking `df` outputs might have a different format (edit: tested in docker, all good)

### CODE REVIEW

**Comment by bard** [comment]
I prefer this approach because it's independent from the deployment platform, but just for completeness, I should mention that Fly allows running multiple processes per machine, so in theory one could run any usual monitoring (disk or otherwise) daemon.

**Comment by boudra** [comment]
> I prefer this approach because it's independent from the deployment platform, but just for completeness, I should mention that Fly allows running multiple processes per machine, so in theory one could run any usual monitoring (disk or otherwise) daemon.

if you mean the ones you can configure in the toml under `[processes]` (process groups), each one runs in it's own machine and cannot share volumes

there's this but as far as i know you can only use it through the machines HTTP API: https://community.fly.io/t/multi-process-machines/8375/15

but yeah this in-process approach should work!

**Comment by bard** [comment]
> if you mean the ones you can configure in the toml under `[processes]` (process groups), each one runs in it's own machine and cannot share volumes

Right, I re-read the relevant docs and it turns out I got it wrong (and Fly has an interesting concept of "process").

**Comment by boudra** [comment]
> > if you mean the ones you can configure in the toml under `[processes]` (process groups), each one runs in it's own machine and cannot share volumes
> 
> Right, I re-read the relevant docs and it turns out I got it wrong (and Fly has an interesting concept of "process").

Yeah exactly 🤣  it's really weird

**Review by vacekj — APPROVED** [review]

File: src/resourceMonitor.test.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
+import { vi, expect, test, describe } from "vitest";
+
+import { createResourceMonitor } from "./resourceMonitor.js";
+
+vi.mock("diskstats", () => {
```
[the remark this answers is not in the document]
Comment (bard) [reply src/resourceMonitor.test.ts:5]: (Likewise for a `logger` dependency actually.)

File: src/resourceMonitor.test.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+      expect(log).toHaveBeenCalledWith({
+        type: "inode",
+        directory: "/",
+        total: 100,
+        used: 60,
+      });
+    }
+
+    // no reporting should happen after the monitor is stopped
```
Comment (bard) [inline src/resourceMonitor.test.ts:76]: Mostly nitpicking here, but would it be difficult (or annoying — read: lots of setup) to have these as individual tests?

**Review by bard — APPROVED** [review]

File: src/config.ts [hunk 3]
```diff
@@ -689,6 +690,11 @@ export function getConfig(): Config {
     .default(null)
     .parse(process.env.BUILD_TAG);
 
+  const enableResourceMonitor = z.coerce
+    .boolean()
+    .default(false)
+    .parse(process.env.ENABLE_RESOURCE_MONITOR);
```
[the remark this answers is not in the document]
Comment (boudra) [reply src/config.ts:696]: damn 🤯

**Review by vacekj — APPROVED** [review]

**Review by bard — APPROVED** [review]
Nice. :+1:

**Pull request merged**